crates/block-producer/src/mempool/batch_graph.rs

```diff
@@ -0,0 +1,209 @@
+        let mut removed = BTreeMap::new();
+
+        let mut to_process = batches;
+
+        while let Some(node_id) = to_process.pop_first() {
+            // Its possible for a node to already have been removed as part of this subgraph
+            // removal.
+            let Some(node) = self.nodes.remove(&node_id) else {
+                continue;
+            };
+
+            // All the child batches are also removed so no need to check for new roots. No new
+            // roots are possible as a result of this subgraph removal.
+            self.roots.remove(&node_id);
+
+            for transaction in &node.transactions {
+                self.transactions.remove(transaction);
+            }
+
+            // Inform parent that this child no longer exists.
+            //
+            // The same is not required for children of this batch as we will be removing those as
+            // well.
+            for parent in &node.parents {
+                // Parent could already be removed as part of this subgraph removal.
+                if let Some(parent) = self.nodes.get_mut(parent) {
+                    parent.children.remove(&node_id);
+                }
+            }
+
+            to_process.extend(node.children);
+            removed.insert(node_id, node.transactions);
+        }
+
+        removed
+    }
+
+    /// Removes a set of batches from the graph without removing any descendants.
+    ///
+    /// This is intended to cull completed batches from stale blocs.
+    pub fn remove_committed(&mut self, batches: BTreeSet<BatchJobId>) -> Vec<TransactionId> {
+        let mut transactions = Vec::new();
+
+        for batch in batches {
+            let node = self.nodes.remove(&batch).expect("Node must be in graph");
+            assert_eq!(node.status, Status::InBlock);
+
+            // Remove batch from graph. No need to update parents as they should be removed in this
+            // call as well.
+            for child in node.children {
+                // Its possible for the child to part of this same set of batches and therefore
+                // already removed.
+                if let Some(child) = self.nodes.get_mut(&child) {
+                    child.parents.remove(&batch);
+                }
+            }
+
+            transactions.extend_from_slice(&node.transactions);
+        }
+
+        transactions
+    }
+
+    /// Mark a batch as proven if it exists.
+    pub fn mark_proven(&mut self, id: BatchJobId, batch: TransactionBatch) {
+        // Its possible for inflight batches to have been removed as part of another batches
+        // failure.
+        if let Some(node) = self.nodes.get_mut(&id) {
+            node.status = Status::Proven(batch);
+            self.try_make_root(id);
+        }
+    }
+
+    /// Returns at most `count` __indepedent__ batches which are ready for inclusion in a block.
+    pub fn select_block(&mut self, count: usize) -> BTreeMap<BatchJobId, TransactionBatch> {
```

Comment:
Would be good to clarify what "independent" means here? I'm assuming that it means no batch depends on another. If so, why do we need this restriction? (i.e., why not return batches with internal dependencies?)

Reply:
I thought you'd specified this 😅 Maybe I misunderstood.

But yes your interpretation is correct; I meant that these batches are not coupled to each other at all.

Reply:
#519